### ISSUE TITLE
Standard way to find modules

### ISSUE THREAD

**Issue by knownasilya** [spolu/breach_core#97]
If I publish a module, how will others find it and know that it's a breach module?
If breach uses npm, is there a specific keyword that should be used?

**Comment by jedahan** [comment]
how does atom.io do it?

**Comment by naknut** [comment]
Thay have their own package manager, its called APM.

**Comment by jacobsmith** [comment]
I would love this as an option -- even just a central area on breach.cc where people could submit modules and then show download stats would be awesome!

**Comment by knownasilya** [comment]
Or even just a list of modules taken by keyword, like [yeoman does](http://yeoman.io/generators/community.html). Pulled from npm or github.

**Comment by spolu** [comment]
So there's a parallel discussion on `mod_strip` I'm going to close the issue there.

**Comment by spolu** [comment]
This could be useful in the meantime: https://github.com/breach/breach_core/wiki/List-of-modules

Someone mentioned tagging on npm and searching by tag within Breach. That could be a good early solution?

**Comment by jedahan** [comment]
+1 on the tagging for now no need to implement yet another package manager especially if most modules are gonna be released via npm anyways...

**Comment by knownasilya** [comment]
@jedahan totally with you on this! Tired of all the package managers.. (component, bower, npm, jspm, apm, meteor, etc...)

**Comment by knownasilya** [comment]
@spolu would love for the breach site to be on github so that others could contribute.

**Comment by spolu** [comment]
@knownasilya it is. `gh-pages` branch

**Comment by spolu** [comment]
Anybody want to take a stance a building such a search engine within breach_core's module manager?

**Comment by knownasilya** [comment]
@spolu :+1:

**Comment by knownasilya** [comment]
> Anybody want to take a stance a building such a search engine within breach_core's module manager?

@spolu will definitely have a look after work.

**Comment by spolu** [comment]
@knownasilya cool!

**Comment by knownasilya** [comment]
@spolu although, would love any direction/guidance as how to start. First time with breach.

**Comment by spolu** [comment]
Of course, and working on breach_core is slightly more difficult than working on a module...
Feel free to ping me on IRC whenever you're ready!

**Comment by miketheprogrammer** [comment]
@spolu I would caution against having a module management system in core. While I hope this project gains amazing popularity it probably wont reach the levels of npm. However, if it should get crazy, decoupling the package manager from the core would prove difficult later on. It is better to start with it decoupled.
Asking the user to download it, similar to mod_strip, when they first open it. Or just automatically downloading it anyway.

**When in doubt, Node Philosophy, Small tight core, large userland.

**Comment by spolu** [comment]
@miketheprogrammer Thanks. That makes sense!

**Comment by knownasilya** [comment]
@miketheprogrammer good insight! This will avoid having to push core updates when the module search needs updating..

**Comment by knownasilya** [comment]
What's a good name? mod_findmod?

**Comment by sbruchmann** [comment]
@knownasilya

Some random suggestions:
- mod_modsearch
- mod_modmngr
- mod_modfinder
- mod_modexplorer

**Comment by knownasilya** [comment]
Thanks @sbruchmann! I've started a repo [here](https://github.com/knownasilya/mod_findmod). It can always be renamed if need be.

**Comment by knownasilya** [comment]
What keyword should be searched, maybe not a keyword at all, just a dependency on `breach_module`? Should we only search npm? Or github as well (projects with a package.json)?

**Comment by spolu** [comment]
Let's start with npm search and see how it goes. People can always add github modules directly through their URL in the meantime.

I'm wondering if we should move the entire module manager there or just the discovery?

**Comment by miketheprogrammer** [comment]
Would recommend discovery and module manager be seperate. Module manager
should be core but discovery should be interchangeable modules.

**Comment by jimmed** [comment]
I've [forked](http://github.com/jimmed/mod_findmod) @knownasilya 's repo and made an initial attempt to call `npm search` from inside the module. However, I've come up against a wall, and it's late at night so I'm leaving it for now. Some details of the problem are commented in [index.js](https://github.com/jimmed/mod_findmod/blob/master/index.js#L17..28), but the best way to experience it is to install the module and see the output for yourselves ;)

**Comment by knownasilya** [comment]
~~@jimmed I installed it into my copy of breach, and don't even get a console log..~~

Get it now, using 'out', but yeah can't find what the issue is.. Maybe `npm.load` needs a different config.

@spolu is there a better way to debug modules, other then using 'out'?

**Comment by corysimmons** [comment]
+1million on some sort of Sublime Text package control replica.

**Comment by knownasilya** [comment]
@corysimmons In what sense? Could you expound just a bit about what you'd like to see.

**Comment by corysimmons** [comment]
![](http://i.imgur.com/udjp0mc.png)

Just make the install modules thing a fuzzy search

**Comment by knownasilya** [comment]
Development has been moved [here](https://github.com/mblarsen/breach-module-manager) and wiki is updated. Anyone looking to help should submit PRs there.

**Comment by spolu** [comment]
Awesome. Keeping the issue open for people to find it and get redirected to your repo.
